proposals/0000-sum-to-tag.md

```diff
@@ -0,0 +1,259 @@
+---
+author: David Feuer
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/369).
+
+# Add `sumToTag#` primop
+
+Add a `sumToTag#` primop indicating which sum alternative is
+represented by a particular unboxed sum value. For example,
+
+```haskell
+sumToTag# (# a | #) = 0#
+sumToTag# (# | | b #) = 2#
+```
+
+## Motivation
+
+It's sometimes useful to get the numerical value of an alternative.
+For example, I might write
+
+```haskell
+newtype Trit# = Trit# (# (##) | (##) | (##) #)
+
+tritToInt# :: Trit# -> Int#
+tritToInt# (Trit# (# (##)|| #)) = 0#
+tritToInt# (Trit# (# |(##)| #)) = 1#
+tritToInt# (Trit# (# ||(##) #)) = 2#
+```
+
+Under the hood, unboxed sums are actually represented by `Int#` tags
+alongside the "payload" values, often with some padding. For example,
+`(# (##) | Bool #)` is represented as something like `(# Int#, Bool #)`.
+The values look like this:
+
+```haskell
+(# (##) | #)   ~=   (# 1#, padding #)
+(# | b #)      ~=   (# 2#, b #)
+```
+
+Note: the choice to make the first alternative `1#` rather than `0#`
+is poorly motivated, and should probably be changed, but that's beyond
+the strict scope of this proposal.
+
+If we could only get our hands on the `Int#` tag, then we could calculate
+`tritToInt#` very efficiently. The C-- code would look much like the
+code that would be produced by
+
+```haskell
+tritToInt'# :: (# Int# #) -> Int#
+tritToInt'# (# t #) = t - 1
+```
+
+Unfortunately, that's *not* the sort of code that `tritToInt#` will
+actually produce. Rather, it will produce code like this:
+
+```haskell
+tritToInt''# :: (# Int# #) -> Int#
+tritToInt''# (# 3# #) -> 2#
+tritToInt''# (# 2# #) -> 1#
+tritToInt''# _ -> 0#
+```
+
+This seems pretty silly, considering that the result of `tritToInt#`
+is likely to be fed straight into some numerical computation. There
+will be extra code and unnecessary conditional branches when all that's
+needed is a simple decrement operation.
+
+## Proposed Change Specification
+
+Add a `sumToTag#` inline primop, analogous to `dataToTag#`. Approximately,
+
+```haskell
+sumToTag# :: forall (xs :: [RuntimeRep]) (a :: TYPE ('SumRep xs)). a -> Int#
+```
+
+`sumToTag#` will return the 0-based index of the sum alternative of its
+argument.
+
+The `forall` feels like a bit of a lie: much like `tagToEnum#` can be compiled
+only when the *type constructor* of its argument is known and is an algebraic
+datatype, `sumToTag#` can be compiled only when the *runtime representation* of
+its argument is known and is `'SumRep xs` for some known `xs`. This restriction
+doesn't actually have any effect in practice, because bindings can't be
+representation-polymorphic anyway.
```

Comment:
```suggestion
It may seem bizarre to have a function that is polymorphic in the runtime representation of its argument; we could only call such a function when we know the concrete representation of its argument, i.e. when the `xs :: [RuntimeRep]` type argument is *closed*. However, GHC already imposes a universal ban on levity-polymorphic arguments (see Section 5.1 of [Levity Polymorphism][]) so we do not have to worry about this case.

[Levity Polymorphism]: https://www.microsoft.com/en-us/research/publication/levity-polymorphism/
```